### PR TITLE
Development

### PR DESCRIPTION
**Pull request by M-Davenport**
merging development branch into master branch

### CODE REVIEW

File: CMP1903_A1_2324/Die.cs [hunk 1]
```diff
@@ -4,20 +4,22 @@
...
 {
-    internal class Die
+    internal class Die //declares the die class
     {
-        /*
-         * The Die class should contain one property to hold the current die value,
-         * and one method that rolls the die, returns and integer and takes no parameters.
-         */
+        private static Random random = new Random(); //declares and initialises a private field for generating random numbers
+
```
Comment (David-Dallakyan) [inline CMP1903_A1_2324/Die.cs:12]: Good use of encapsulation here

File: CMP1903_A1_2324/Die.cs [hunk 2]
```diff
@@ -4,20 +4,22 @@
...
-namespace CMP1903_A1_2324
+namespace CMP1903_A1_2324 //defines the namespace
 {
-    internal class Die
+    internal class Die //declares the die class
     {
-        /*
-         * The Die class should contain one property to hold the current die value,
-         * and one method that rolls the die, returns and integer and takes no parameters.
-         */
```
Comment (David-Dallakyan) [inline CMP1903_A1_2324/Die.cs:14]: Appropriately labelling classes, good use of the 'Random' class.

File: CMP1903_A1_2324/Die.cs [hunk 3]
```diff
@@ -4,20 +4,22 @@
...
+        private static Random random = new Random(); //declares and initialises a private field for generating random numbers
+        
+        public int FaceValue { get; private set; } //declares a property to hold the current die value
 
-        //Property
-
-
-        //Method
+        public int Roll() //declares a method to roll the die
+        {
```
Comment (David-Dallakyan) [inline CMP1903_A1_2324/Die.cs:16]: Could explain in the comment the reason behind the range of 1-6 for the die rolls

File: CMP1903_A1_2324/Game.cs [hunk 4]
```diff
@@ -1,21 +1,32 @@
 ﻿using System;
```
Comment (David-Dallakyan) [inline CMP1903_A1_2324/Game.cs:1]: Good use of encapsulation and instantiation used here for the die objects.

File: CMP1903_A1_2324/Game.cs [hunk 5]
```diff
@@ -1,21 +1,32 @@
...
-using System.Collections.Generic;
```
Comment (David-Dallakyan) [inline CMP1903_A1_2324/Game.cs:2]: Comments can be understood easily, used well

File: CMP1903_A1_2324/Game.cs [hunk 6]
```diff
@@ -1,21 +1,32 @@
...
-         */
+        //declares and initialises private fields for each die
+        private Die die1 = new Die();
+        private Die die2 = new Die();
+        private Die die3 = new Die();
 
-        //Methods
+        public void PlayGame() //declares the PlayGame method
+        {
+            //rolls the dice three times and stores the results
```
Comment (David-Dallakyan) [inline CMP1903_A1_2324/Game.cs:14]: Could use error handling here, in case the Roll method were to fail.

File: CMP1903_A1_2324/Program.cs [hunk 7]
```diff
@@ -1,19 +1,17 @@
...
-using System.Linq;
-using System.Text;
-using System.Threading.Tasks;
 
-namespace CMP1903_A1_2324
+namespace CMP1903_A1_2324//defines the namespace
 {
-    internal class Program
+    internal class Program//declares the class called Program
     {
```
Comment (David-Dallakyan) [inline CMP1903_A1_2324/Program.cs:6]: Clear and consise use of the main method

File: CMP1903_A1_2324/Program.cs [hunk 8]
```diff
@@ -1,19 +1,17 @@
...
-        static void Main(string[] args)
+        static void Main(string[] args) //declares the main method
         {
-            /*
```
Comment (David-Dallakyan) [inline CMP1903_A1_2324/Program.cs:13]: Good instantiation used on 'game' and 'testing' objects

File: CMP1903_A1_2324/Testing.cs [hunk 9]
```diff
@@ -1,20 +1,23 @@
 ﻿using System;
```
Comment (David-Dallakyan) [inline CMP1903_A1_2324/Testing.cs:1]: The testing methods are well commented

File: CMP1903_A1_2324/Testing.cs [hunk 10]
```diff
@@ -1,20 +1,23 @@
...
-        //Method
+            //tests the Die class
+            Die die = new Die(); //creates new instance of the die class to test if rolls are between 1 and 6
+            int rolledValue = die.Roll(); //calls the roll method and stores the value of the roll
+            //asserts that the rolled value is be between 1 and 6. if not an error message is produced
+            Debug.Assert(rolledValue >= 1 && rolledValue <= 6, "Rolled value is not between 1 and 6.");
+        }
     }
 }
+
```
Comment (David-Dallakyan) [inline CMP1903_A1_2324/Testing.cs:23]: Ultimately, your code is well structured and you've shown a good understanding of instantiation and encapsulation.  To improve, you can consider making more detailed comments and include error handling.

**Review by David-Dallakyan — COMMENTED** [review]
To summarise, your code is well structured and you've shown a good understanding of instantiation and encapsulation. To improve, you can consider making more detailed comments and include error handling.

File: CMP1903_A1_2324/Die.cs [hunk 11]
```diff
@@ -4,20 +4,22 @@
...
-    internal class Die
+    internal class Die //declares the die class
     {
-        /*
-         * The Die class should contain one property to hold the current die value,
-         * and one method that rolls the die, returns and integer and takes no parameters.
-         */
+        private static Random random = new Random(); //declares and initialises a private field for generating random numbers
+        
+        public int FaceValue { get; private set; } //declares a property to hold the current die value
```
Comment (eth23git) [inline CMP1903_A1_2324/Die.cs:13]: Encapsulation used well even on public variable allowing it to be read publicly but only changed within the class.

File: CMP1903_A1_2324/Game.cs [hunk 12]
```diff
@@ -1,21 +1,32 @@
...
+            Console.WriteLine("Die 2 rolled: " + roll2);
+            Console.WriteLine("Die 3 rolled: " + roll3);
 
+            int total = roll1 + roll2 + roll3; //calculates the total of the dice rolls
+            Console.WriteLine("Total of three rolls: " + total);//prints the total of the dice rolls
+        }
+
+        public int GetTotal() //method to get the total of the three dice rolls
+        {
+            return die1.FaceValue + die2.FaceValue + die3.FaceValue; //returns the sum of the face values of all three dice
```
Comment (eth23git) [inline CMP1903_A1_2324/Game.cs:29]: Could be more concise to sum roll1, roll2 and roll3 instead, allowing FaceValue to be fully encapsulated as these are already in the same class.

File: CMP1903_A1_2324/Game.cs [hunk 13]
```diff
@@ -1,21 +1,32 @@
...
+        private Die die2 = new Die();
+        private Die die3 = new Die();
 
-        //Methods
+        public void PlayGame() //declares the PlayGame method
+        {
+            //rolls the dice three times and stores the results
+            int roll1 = die1.Roll();
+            int roll2 = die2.Roll();
+            int roll3 = die3.Roll();
```
Comment (eth23git) [inline CMP1903_A1_2324/Game.cs:17]: Good demonstration of the benefits of instantiation.

File: CMP1903_A1_2324/Testing.cs [hunk 14]
```diff
@@ -1,20 +1,23 @@
...
+            int total = game.GetTotal(); //calls the gettotal method to get the total of the dice rolls
+            //asserts that the total of the three dice rolls is between 3 and 18. If not an error message is produced
+            Debug.Assert(total >= 3 && total <= 18, "Total of three dice rolls is not as between 3 and 18.");
 
-        //Method
+            //tests the Die class
+            Die die = new Die(); //creates new instance of the die class to test if rolls are between 1 and 6
+            int rolledValue = die.Roll(); //calls the roll method and stores the value of the roll
+            //asserts that the rolled value is be between 1 and 6. if not an error message is produced
+            Debug.Assert(rolledValue >= 1 && rolledValue <= 6, "Rolled value is not between 1 and 6.");
```
Comment (eth23git) [inline CMP1903_A1_2324/Testing.cs:19]: Effective testing of the die class and its method Roll().

File: CMP1903_A1_2324/Testing.cs [hunk 15]
```diff
@@ -1,20 +1,23 @@
...
-         * Create a Game object, call the methods and compare their output to expected output.
-         * Create a Die object and call its method.
-         * Use debug.assert() to make the comparisons and tests.
-         */
+        public void TestGameAndDie(Game game) //declares the TestGameAndDie method and takes a Game object as a parameter
+        {
+            //tests the Game class            
+            int total = game.GetTotal(); //calls the gettotal method to get the total of the dice rolls
+            //asserts that the total of the three dice rolls is between 3 and 18. If not an error message is produced
+            Debug.Assert(total >= 3 && total <= 18, "Total of three dice rolls is not as between 3 and 18.");
```
Comment (eth23git) [inline CMP1903_A1_2324/Testing.cs:13]: Good way of debugging, ensuring min and max value of total I hadn't thought of.

File: CMP1903_A1_2324/Program.cs [hunk 16]
```diff
@@ -1,19 +1,17 @@
...
-             * Create a Game object and call its methods.
-             * Create a Testing object to verify the output and operation of the other classes.
-             */
+            Game game = new Game(); //creates a Game object
+            game.PlayGame(); //calls the PlayGame() method of the Game object
+
+            Testing testing = new Testing(); //initialises the testing object
+            testing.TestGameAndDie(game); //calls the method and passes the game object to the TestGameAndDie() method
```
Comment (eth23git) [inline CMP1903_A1_2324/Program.cs:13]: Good commenting additions from previous version allow for a very easy understanding.

File: CMP1903_A1_2324/Testing.cs (hunk 10 above)
Comment (eth23git) [inline CMP1903_A1_2324/Testing.cs:23]: The code is very well commented making it very simple to understand and comment on, mostly effectively utilising encapsulation and instantiation. Requirements are met well and code is concise to achieve the requirements. There is little evidence to show the style guidelines have been adhered to and there is effective use of error handling.

**Pull request merged**